### PR TITLE
feat: support etcd cluster password authentication

### LINKED ISSUE #100: support etcd cluster inspection,which is password authentication
So far, kstone can't inspect etcd cluster which is password authentication

### PR DESCRIPTION
**Pull request by lianghao208**
/cc @tangcong @engow 
fix #100

### CODE REVIEW

**Comment by jianhaiqing** [comment]
cool to see the feature

File: pkg/clusterprovider/helper.go [hunk 1]
```diff
@@ -71,16 +70,21 @@ func populateExtensionCientURLMap(extensionClientURLs string) (map[string]string
...
+	cli *etcd.Client) ([]kstonev1alpha2.MemberStatus, error) {
 	etcdMembers := make([]kstonev1alpha2.MemberStatus, 0)
 
-	ca, cert, key := "", "", ""
-	if tls != nil {
-		ca, cert, key = tls.TrustedCAFile, tls.CertFile, tls.KeyFile
+	config := &etcd.ClientConfig{
+		Endpoints: endpoints,
+	}
+	if cli != nil {
```
Comment (tangcong) [inline pkg/clusterprovider/helper.go:79]: if cli != nil && cli.TLS != nil {

File: pkg/clusterprovider/helper.go [hunk 2]
```diff
@@ -191,15 +195,20 @@ func GetEtcdClusterMemberStatus(
...
+	cli *etcd.Client) ([]EtcdAlarm, error) {
 	etcdAlarms := make([]EtcdAlarm, 0)
 
-	ca, cert, key := "", "", ""
-	if tls != nil {
-		ca, cert, key = tls.TrustedCAFile, tls.CertFile, tls.KeyFile
+	config := &etcd.ClientConfig{
+		Endpoints: endpoints,
+	}
+	if cli != nil {
```
Comment (tangcong) [inline pkg/clusterprovider/helper.go:204]: +1

File: pkg/etcd/client.go [hunk 3]
```diff
@@ -33,31 +33,37 @@ import (
...
+type ClientSecret struct {
 	kubeCli      kubernetes.Interface
 	secretLister listerscorev1.SecretLister
-	secretGetter func(t *TLSSecret, namespace, secretName string) (*v1.Secret, error)
+	secretGetter func(t *ClientSecret, namespace, secretName string) (*v1.Secret, error)
 }
 
-func NewTLSSecretGetter(clientBuilder util.ClientBuilder) *TLSSecret {
-	return &TLSSecret{
+type Client struct {
```
Comment (tangcong) [inline pkg/etcd/client.go:46]: Client -> ClientConfig

File: pkg/etcd/client.go [hunk 4]
```diff
@@ -33,31 +33,37 @@ import (
...
+	Username string
+	Password string
+	TLS      *transport.TLSInfo
+}
+
+func NewClientSecretGetter(clientBuilder util.ClientBuilder) *ClientSecret {
```
Comment (tangcong) [inline pkg/etcd/client.go:52]: NewClientSecretGetter -> NewClientConfigFromSecret

File: pkg/etcd/client.go [hunk 5]
```diff
@@ -33,31 +33,37 @@ import (
...
+	return &ClientSecret{
 		kubeCli:      clientBuilder.ClientOrDie(),
 		secretGetter: Secret,
 	}
 }
 
-func NewTLSSecretCacheGetter(secretLister listerscorev1.SecretLister) *TLSSecret {
-	return &TLSSecret{
+func NewClientSecretCacheGetter(secretLister listerscorev1.SecretLister) *ClientSecret {
```
Comment (tangcong) [inline pkg/etcd/client.go:59]: NewClientSecretCacheGetter -> NewClientFromSecretCache

File: pkg/etcd/client.go [hunk 6]
```diff
@@ -33,31 +33,37 @@ import (
...
 )
 
-type TLSGetter interface {
-	Config(path string, sc string) (*transport.TLSInfo, error)
+type ClientGetter interface {
+	Config(path string, sc string) (*Client, error)
 }
 
-type TLSSecret struct {
+type ClientSecret struct {
```
Comment (tangcong) [inline pkg/etcd/client.go:40]: ClientSecret -> ClientConfigSecret

File: pkg/inspection/consistency.go [hunk 7]
```diff
@@ -69,7 +64,19 @@ func (c *Server) getEtcdConsistentMetadata(
 				klog.Errorf("failed to get etcd stat backend,backend %s,err is %v", extensionClientURL, err)
 				return err
 			}
-			err = backend.Init(ca, cert, key, extensionClientURL)
+
+			config := &etcd.ClientConfig{
+				Endpoints: []string{extensionClientURL},
+			}
+
+			if cli != nil {
```
Comment (tangcong) [inline pkg/inspection/consistency.go:72]: +1

File: pkg/inspection/request.go [hunk 8]
```diff
@@ -79,12 +79,18 @@ func (c *Server) CollectEtcdClusterRequest(inspection *kstonev1alpha2.EtcdInspec
...
 
-	ca, cert, key := "", "", ""
-	if tlsConfig != nil {
-		ca, cert, key = tlsConfig.TrustedCAFile, tlsConfig.CertFile, tlsConfig.KeyFile
+	config := &etcd.ClientConfig{
+		Endpoints: clusterprovider.GetStorageMemberEndpoints(cluster),
 	}
 
-	client, err := etcd.NewClientv3(ca, cert, key, clusterprovider.GetStorageMemberEndpoints(cluster))
+	if cli != nil {
```
Comment (tangcong) [inline pkg/inspection/request.go:86]: +1

File: pkg/router/router.go [hunk 9]
```diff
@@ -187,21 +187,28 @@ func EtcdKeyList(ctx *gin.Context) {
...
-	ca, cert, key := "", "", ""
-	if tlsConfig != nil {
-		ca, cert, key = tlsConfig.TrustedCAFile, tlsConfig.CertFile, tlsConfig.KeyFile
+	config := &etcd.ClientConfig{
+		Endpoints: []string{cluster.Status.ServiceName},
 	}
-	klog.Infof("endpoint: %s, ca: %s, cert: %s, key: %s", cluster.Status.ServiceName, ca, cert, key)
-	client, err := etcd.NewClientv3(ca, cert, key, []string{cluster.Status.ServiceName})
+
+	if cli != nil {
```
Comment (tangcong) [inline pkg/router/router.go:203]: +1

File: pkg/etcd/client.go [hunk 10]
```diff
@@ -33,31 +33,37 @@ import (
 	"tkestack.io/kstone/pkg/controllers/util"
 )
 
-type TLSGetter interface {
-	Config(path string, sc string) (*transport.TLSInfo, error)
+type ClientGetter interface {
```
Comment (tangcong) [inline pkg/etcd/client.go:36]: ClientGetter -> ClientConfigGetter

File: pkg/etcd/client.go [hunk 11]
```diff
@@ -33,31 +33,37 @@ import (
...
+	Config(path string, sc string) (*Client, error)
```
Comment (tangcong) [inline pkg/etcd/client.go:37]: Config -> New or Load

**Review by tangcong — APPROVED** [review]
lgtm. good. thanks.

**Pull request merged**